go/apps/gw/services/auth/auth.go

```diff
@@ -0,0 +1,167 @@
+package auth
+
+import (
+	"context"
+	"strings"
+
+	"github.com/unkeyed/unkey/go/apps/gw/server"
+	partitionv1 "github.com/unkeyed/unkey/go/gen/proto/partition/v1"
+	"github.com/unkeyed/unkey/go/internal/services/keys"
+	"github.com/unkeyed/unkey/go/pkg/assert"
+	"github.com/unkeyed/unkey/go/pkg/codes"
+	"github.com/unkeyed/unkey/go/pkg/fault"
+	"github.com/unkeyed/unkey/go/pkg/otel/logging"
+	"github.com/unkeyed/unkey/go/pkg/zen"
+)
+
+type authenticator struct {
+	logger logging.Logger
+	keys   keys.KeyService
+}
+
+var _ Authenticator = (*authenticator)(nil)
+
+// New creates a new authenticator with the given configuration.
+func New(config Config) (Authenticator, error) {
+	if err := assert.All(
+		assert.NotNilAndNotZero(config.Logger, "Logger is required"),
+		assert.NotNilAndNotZero(config.Keys, "Keys DB is required"),
+	); err != nil {
+		return nil, err
+	}
+
+	return &authenticator{
+		logger: config.Logger,
+		keys:   config.Keys,
+	}, nil
+}
+
+// Authenticate processes API key authentication for the request.
+func (a *authenticator) Authenticate(ctx context.Context, sess *server.Session, config *partitionv1.GatewayConfig) error {
+	// Skip authentication if not configured or not enabled
+	if config.AuthConfig == nil || !config.AuthConfig.Enabled || !config.AuthConfig.RequireApiKey {
+		return nil
+	}
+
+	// Extract API key from Authorization header
+	apiKey, err := a.extractAPIKey(sess)
+	if err != nil {
+		return err
+	}
+
+	// Verify the API key
+	return a.verifyAPIKey(ctx, sess, apiKey, config)
+}
+
+// extractAPIKey extracts and validates the API key from the Authorization header.
+func (a *authenticator) extractAPIKey(sess *server.Session) (string, error) {
+	req := sess.Request()
+
+	// Check for Authorization header
+	authHeader := req.Header.Get("Authorization")
+	if authHeader == "" {
+		return "", fault.New("missing authorization header",
+			fault.Code(codes.Gateway.Auth.Unauthorized.URN()),
+			fault.Public("Authorization header required"),
+		)
+	}
+
+	// Validate Bearer token format
+	if !strings.HasPrefix(authHeader, "Bearer ") {
+		return "", fault.New("invalid authorization header format",
+			fault.Code(codes.Gateway.Auth.Unauthorized.URN()),
+			fault.Public("Invalid authorization header format"),
+		)
+	}
+
+	// Extract API key
+	apiKey := strings.TrimPrefix(authHeader, "Bearer ")
+	if apiKey == "" {
+		return "", fault.New("empty api key",
+			fault.Code(codes.Gateway.Auth.Unauthorized.URN()),
+			fault.Public("API key is required"),
+		)
+	}
+
+	return apiKey, nil
+}
+
+// verifyAPIKey retrieves and validates the API key against the configured keyspace.
+func (a *authenticator) verifyAPIKey(ctx context.Context, sess *server.Session, apiKey string, config *partitionv1.GatewayConfig) error {
+	z := zen.Session{
```

Comment:
this looks like a mistake, the gateway has its own session code, why are we now using zen sessions?